nan.h

```diff
@@ -1852,75 +1852,47 @@ NAN_INLINE void SetInstanceTemplate(
 }
 
-#if NODE_MODULE_VERSION < IOJS_3_0_MODULE_VERSION
-NAN_INLINE void SetMethod(
-    v8::Handle<v8::Object> recv
-  , const char *name
-  , FunctionCallback callback) {
-  HandleScope scope;
-  v8::Local<v8::Function> fn = GetFunction(New<v8::FunctionTemplate>(
-      callback)).ToLocalChecked();
-  v8::Local<v8::String> fn_name = New(name).ToLocalChecked();
-  fn->SetName(fn_name);
-  recv->Set(fn_name, fn);
-}
+namespace imp {
 
-NAN_INLINE void SetMethod(
-    v8::Handle<v8::FunctionTemplate> templ
-  , const char *name
-  , FunctionCallback callback) {
-  HandleScope scope;
-  v8::Local<v8::FunctionTemplate> t = New<v8::FunctionTemplate>(callback);
-  v8::Local<v8::String> fn_name = New(name).ToLocalChecked();
-  t->SetClassName(fn_name);
-  templ->Set(fn_name, t);
+// Note(@agnat): Helper to distinguish different receiver types. The first
+// version deals with receivers derived from v8::Template. The second version
+// handles everything else. The final argument only serves as discriminator and
+// is unused.
+template <typename T>
+NAN_INLINE
+void
+SetMethodAux(T recv,
+             v8::Local<v8::String> name,
+             v8::Local<v8::FunctionTemplate> tpl,
+             v8::Template *) {
+  recv->Set(name, tpl);
 }
 
-NAN_INLINE void SetMethod(
-    v8::Handle<v8::ObjectTemplate> templ
-  , const char *name
-  , FunctionCallback callback) {
-  HandleScope scope;
-  v8::Local<v8::FunctionTemplate> t = New<v8::FunctionTemplate>(callback);
-  v8::Local<v8::String> fn_name = New(name).ToLocalChecked();
-  t->SetClassName(fn_name);
-  templ->Set(fn_name, t);
-}
-#else
-NAN_INLINE void SetMethod(
-    v8::Local<v8::Object> recv
-  , const char *name
-  , FunctionCallback callback) {
-  HandleScope scope;
-  v8::Local<v8::Function> fn = GetFunction(New<v8::FunctionTemplate>(
-      callback)).ToLocalChecked();
-  v8::Local<v8::String> fn_name = New(name).ToLocalChecked();
-  fn->SetName(fn_name);
-  recv->Set(fn_name, fn);
+template <typename T>
+NAN_INLINE
+void
+SetMethodAux(T recv,
+             v8::Local<v8::String> name,
+             v8::Local<v8::FunctionTemplate> tpl,
+             ...) {
+  recv->Set(name, GetFunction(tpl).ToLocalChecked());
 }
 
-NAN_INLINE void SetMethod(
-    v8::Local<v8::FunctionTemplate> templ
-  , const char *name
-  , FunctionCallback callback) {
-  HandleScope scope;
-  v8::Local<v8::FunctionTemplate> t = New<v8::FunctionTemplate>(callback);
-  v8::Local<v8::String> fn_name = New(name).ToLocalChecked();
-  t->SetClassName(fn_name);
-  templ->Set(fn_name, t);
-}
+} // end of namespace imp
 
+template <typename T, template <typename> class HandleType>
 NAN_INLINE void SetMethod(
-    v8::Local<v8::ObjectTemplate> templ
+    HandleType<T> recv
   , const char *name
   , FunctionCallback callback) {
   HandleScope scope;
-  v8::Local<v8::FunctionTemplate> t = New<v8::FunctionTemplate>(callback);
+  v8::Local<v8::FunctionTemplate> tpl = New<v8::FunctionTemplate>(callback);
```

Comment:
Yeah, the diff is all over the place. Not sure why though. I didn't rename the local. Diff just *thinks* I did. :-/

I'll see what I can do.